### DISCUSSION TITLE
[mini.surround] has no effect, as if `require('mini.surround').setup()` wasn't there

### DISCUSSION THREAD

**Discussion by proofer** [nvim-mini/mini.nvim#1193] [Q&A] [answered]
I'm stumped. The title's setup() call is among those of several other mini.nvim sub-plugins within the lazy.nvim installation of the mini.nvim library;' all of of the others work. But if I type `s` in normal mode, the character under the cursor is deleted and insert mode is entered  -- the standard vim action.  `which-key` shows no mappings starting with `s`.  Any ideas for chasing this down?

neovim 0.10.1; macOS

**Comment by abeldekat** [comment]
When you enter `:Lazy`: Is the plugin in the `loaded` section? Next to the plugin, the keys in use are displayed.

A scenario I can think of when using `LazyVim`:

1. Extra `mini.surround` enabled. Default keys start with `gs`
2. Flash disabled

**Reply by proofer** [reply]
I'm using the lazy.nvim plugin manager but not LazyVim. `:Lazy` shows mini.nvim in the `Loaded` list, with no keys listed. It shows no info about the component modules of mini.nvim. There is no extra `mini.surround` configured. If by Flash you mean Folkes's flash.nvim plugin, I have never installed it. 
<img width="302" alt="image" src="https://github.com/user-attachments/assets/9d0e1a76-7a55-4e48-90a2-33c4c5c1240f">

**Reply by abeldekat** [reply]
Could you share the (relevant parts of) the config?

**Reply by proofer** [reply]
```
return {
    'echasnovski/mini.nvim',
    config = function()
        local map = function(keys, func, desc)
            vim.keymap.set('n', keys, func, { desc = desc })
        end
        --
        -- Better Around/Inside textobjects, e.g.:
        --  - va)  - [V]isually select [A]round [)]paren
        --  - yinq - [Y]ank [I]nside [N]ext [']quote
        --  - ci'  - [C]hange [I]nside [']quote
        require('mini.ai').setup({
            n_lines = 128, -- text object search range
        })

        require('mini.files').setup({
            map('<leader>m', function()
                MiniFiles.open(MiniFiles.get_latest_path())
            end, 'Open mini.files in parent directory'),
        })

        require('mini.indentscope').setup({
            draw = {
                animation = function()
                    return 0
                end,
            },
        })

        require('mini.jump2d').setup({
            map('f', function()
                MiniJump2d.start(MiniJump2d.builtin_opts.single_character)
            end, 'Mini.jump2d to specified character'),
        })

        require('mini.pairs').setup()

        -- Add/delete/replace surroundings (brackets, quotes, etc.)
        --
        -- - saiw) - [S]urround [A]dd [I]nner [W]ord [)]Paren
        -- - sd'   - [S]urround [D]elete [']quotes
        -- - sr)'  - [S]urround [R]eplace [)] [']
        require('mini.surround').setup()

        require('mini.sessions').setup()

        require('mini.misc').setup()
        MiniMisc.setup_auto_root()

        --
    end,
}

```

**Reply by abeldekat** [reply]
In that fragment, for files and jump2d, you invoke a custom map function inside the table passed to `mini`. I have never seen that.
What happens if you un-comment al inner setup calls, except for `mini.surround`?

**Reply by abeldekat** [reply]
Using the minimal repro from lazy.nvim. Surround works.

Save as `repro.lua`, run with `nvim -u repro.lua`:

```lua
vim.env.LAZY_STDPATH = ".repro"
load(vim.fn.system("curl -s https://raw.githubusercontent.com/folke/lazy.nvim/main/bootstrap.lua"))()

require("lazy.minit").repro({
  spec = {
    -- add any other plugins here

    {
      "echasnovski/mini.nvim",
      config = function()
        local map = function(keys, func, desc)
          vim.keymap.set("n", keys, func, { desc = desc })
        end
        --
        -- Better Around/Inside textobjects, e.g.:
        --  - va)  - [V]isually select [A]round [)]paren
        --  - yinq - [Y]ank [I]nside [N]ext [']quote
        --  - ci'  - [C]hange [I]nside [']quote
        require("mini.ai").setup({
          n_lines = 128, -- text object search range
        })

        require("mini.files").setup({
          map("<leader>m", function()
            MiniFiles.open(MiniFiles.get_latest_path())
          end, "Open mini.files in parent directory"),
        })

        require("mini.indentscope").setup({
          draw = {
            animation = function()
              return 0
            end,
          },
        })

        require("mini.jump2d").setup({
          map("f", function()
            MiniJump2d.start(MiniJump2d.builtin_opts.single_character)
          end, "Mini.jump2d to specified character"),
        })

        require("mini.pairs").setup()

        -- Add/delete/replace surroundings (brackets, quotes, etc.)
        --
        -- - saiw) - [S]urround [A]dd [I]nner [W]ord [)]Paren
        -- - sd'   - [S]urround [D]elete [']quotes
        -- - sr)'  - [S]urround [R]eplace [)] [']
        require("mini.surround").setup()

        require("mini.sessions").setup()

        require("mini.misc").setup()
        MiniMisc.setup_auto_root()

        --
      end,
    },
  },
})
```

**Reply by abeldekat** [reply]
Perhaps un-comment all plugins loaded after `mini.nvim`, to find the culprit?

**Reply by proofer** [reply]
Please keep in mind that everything works, and has been doing so for many months, *except* mini.surround.  And until recently, mini.surround was working, too. But a couple of months ago I had some issues with plugin interactions and pinned mini.nvim, which-key, and one other plugin to a previous commit. A few days ago I un-pinned them to use current versions.

**Reply by abeldekat** [reply]
It's difficult to tell without seeing your config.

**Reply by proofer** [reply]
https://github.com/proofer/config.nvim

**Reply by abeldekat** [reply]
I get a page-not-found....)

**Reply by abeldekat** [reply]
Is it public?

**Reply by proofer** [reply]
Sorry; fixed.

**Reply by echasnovski** [reply]
> Please keep in mind that everything works, and has been doing so for many months, _except_ mini.surround. And until recently, mini.surround was working, too. But a couple of months ago I had some issues with plugin interactions and pinned mini.nvim, which-key, and one other plugin to a previous commit. A few days ago I un-pinned them to use current versions.

This sounds like an issue with 'which-key.nvim' rewrite about two months. I *think* 'timeoutlen' is the issue and [this should help](https://github.com/echasnovski/mini.nvim/blob/73f2e7a596bbb4e6ae6728d3a4426221ffc9512d/doc/mini-surround.txt#L570-L574). But here is a fuller list of steps I'd do to eliminate basic reasons:
- Execute `:=MiniSurround`. If it prints `nil`, then `require('mini.surround').setup()` was not called. Make sure it is called.
- Execute `:nmap s`. It should show a bunch of mappings from 'mini.surround'. If not, make sure there is no plugin interactive with those keymaps later.
- If both previous steps are resolved, make sure to have high enough 'timeoutlen' (by adding `vim.o.timeoutlen = 1000` in your config) and/or disable built-in `s` key (by adding `vim.keymap.set({ 'n', 'x' }, 's', '<Nop>')` line to your config). I personally use the second approach.

**Reply by abeldekat** [reply]
Thanks! I'll take a look.

**Reply by abeldekat** [reply, marked as answer]
I also think it's a matter of timeoutlen. When I type `sa` in quick succession, surround kicks in.

**Reply by proofer** [reply]
`timeoutlen` is it! Thanks very much for your help!

**Reply by echasnovski** [reply]
@abeldekat, thanks for taking your time to troubleshoot this! Highly appreciated!

**Reply by abeldekat** [reply]
Most welcome!

**Reply by sunbearc22** [reply]
I encountered this same issue too and setting `vim.o.timeoutlen = 1000` solved it. I would like to suggest mentioning this requirement in the `README.md` so that users of `mini.surround` are informed at the get-go.   @proofer  Thank u for raising this issue.

**Reply by echasnovski** [reply]
> I encountered this same issue too and setting `vim.o.timeoutlen = 1000` solved it. I would like to suggest mentioning this requirement in the `README.md` so that users of `mini.surround` are informed at the get-go.

This is not specific to 'mini.surround', but for any mapping that consists from two and more characters.

The `vim.o.timeoutlen = 1000` is the default value, so not setting it initially would be a better idea.

**Reply by sunbearc22** [reply]
@echasnovski Thank you for explaining. In my case, I had installed several mini plugins. However, all the other plugins worked except `mini.surround`. It seems to be more sensitive to this parameter than the other plugins that I had installed. I think it can be helpful to remind/inform users not to use too small a value for `timeoutlen` as it can hinder this plugin from functioning.  Will leave this decision to you all. :wink:
